src/Pack/Runner.idr

```diff
@@ -141,46 +144,50 @@ isFetch _     = False
 export covering
 runCmd : HasIO io => EitherT PackErr io ()
 runCmd = do
-  pd <- getPackDir
-  withTmpDir $ do
-    cd       <- CD <$> curDir
-    cache    <- emptyCache
-    (mc,cmd) <- getConfig Cmd
-    let fetch := isFetch (fst cmd)
-    linebuf  <- getLineBufferingCmd
-    case cmd of
-      (Completion ** [a,b])     => env mc fetch >>= complete a b
-      (CompletionScript ** [f]) => putStrLn (completionScript f)
-      (Query  ** [MkQ m s])     => env mc fetch >>= query m s
-      (Fuzzy ** [MkFQ m s])     => idrisEnv mc fetch >>= fuzzy m s
-      (UpdateDB ** [])          => updateDB
-      (CollectGarbage ** [])    => env mc fetch >>= garbageCollector
-      (Run ** [p,args])         => idrisEnv mc fetch >>= runApp !(refinePkg p) args
-      (Test ** [p,args])        => idrisEnv mc fetch >>= runTest p args
-      (Exec ** [p,args])        => idrisEnv mc fetch >>= exec p args
-      (Repl ** [p])             => idrisEnv mc fetch >>= idrisRepl p
-      (Build ** [p])            => idrisEnv mc fetch >>= build !(refinePkg p)
-      (BuildDeps ** [p])        => idrisEnv mc fetch >>= buildDeps !(refinePkg p)
-      (Typecheck ** [p])        => idrisEnv mc fetch >>= typecheck !(refinePkg p)
-      (Clean ** [p])            => idrisEnv mc fetch >>= clean !(refinePkg p)
-      (CleanBuild ** [p])       => do p <- refinePkg p
-                                      e <- idrisEnv mc fetch
-                                      clean p e >> build p e
-      (PrintHelp ** [c])        => putStrLn (usageDesc c)
-      (Install ** [ps])         => idrisEnv mc fetch >>= \e => installLibs ps
-      (Remove ** [ps])          => idrisEnv mc fetch >>= \e => removeLibs ps
-      (InstallApp ** [ps])      => idrisEnv mc fetch >>= \e => installApps ps
-      (RemoveApp ** [ps])       => idrisEnv mc fetch >>= \e => removeApps ps
-      (Update ** [])            => idrisEnv mc fetch >>= update
-      (Fetch ** [])             => idrisEnv mc fetch >>= \e => install []
-      (PackagePath ** [])       => env mc fetch >>= packagePathDirs >>= putStrLn
-      (LibsPath ** [])          => env mc fetch >>= packageLibDirs  >>= putStrLn
-      (DataPath ** [])          => env mc fetch >>= packageDataDirs >>= putStrLn
-      (AppPath ** [n])          => env mc fetch >>= appPath n
-      (Info ** [])              => env mc fetch >>= printInfo
-      (New ** [pty,p])          => idrisEnv mc fetch >>= new cd pty p
-      (Switch ** [db])          => do
-        env <- idrisEnv mc fetch
-        install []
-        writeCollection
-      (Uninstall ** [])         => uninstallPack @{metaConfigToLogRef @{mc}}
+  pd  <- getPackDir
+  td  <- mkTmpDir
+  liftIO $ syncApp $ ignore $ race
+    [ onSignal SigINT (liftIO $ run $ rmDir tmpDir)
+    , liftIO $ run $ withTmpDir' $ do
```

Comment:
Looks like instead of changing signatures of functions (making `TmpDir` an explicit argument) and doing running the thread in the runner, it's better to do all of this in the original `withTmpDir` function, leaving the rest unchanged. I think, added new functionality is the responsibility of `withTmpDir`.

Reply:
@buzden I really like this idea, I believe I've addressed this via 33745c63861e63c488e0c35b73b19d0025041850.